File: DFS_BFS/boj_2178_미로탐색.java
package DFS_BFS;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;


class Coordinates{
	private int x;
	private int y;
	
	public Coordinates(int x, int y) {
		this.x = x;
		this.y = y;
	}
	public int getX() {
		return x;
	}
	public int getY() {
		return y;
	}
}

public class boj_2178_미로탐색 {
	static int N, M;
	static int[][] maze;
	static int[] dx = { -1, 0 , 1, 0};
	static int[] dy = { 0, -1 , 0, 1};
	static int count = 0;
	
	public static void bfs(int x, int y) {
		Queue<Coordinates> queue = new LinkedList<>();
		
		 queue.offer(new Coordinates(x, y));
		 
		 while(!queue.isEmpty()) {
			 int currX = queue.peek().getX();
			 int currY = queue.poll().getY();
			 
			 for (int i = 0; i < 4; i++) {
				int nextX = currX + dx[i];
				int nextY = currY + dy[i];
				
				if(nextX < 0 || nextY < 0 || nextX > N-1 || nextY > M-1)
					continue;
				if (maze[nextX][nextY] == 1) {
					queue.offer(new Coordinates(nextX, nextY));
					maze[nextX][nextY] = maze[currX][currY]+1;
					
					if(nextX == N-1 && nextY == M-1) {
						return;
					}
				}
			}
		 }
		
	}
	public static void main(String[] args)throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		maze = new int[N][M];
		for (int i = 0; i < N; i++) {
			String str = br.readLine();
			for (int j = 0; j < M; j++) {
				maze[i][j] = str.charAt(j) - '0';
			}
		}
		bfs(0,0);
		System.out.println(maze[N-1][M-1]);
		
	}
}

File: DFS_BFS/boj_2178_미로탐색.cpp
#include <iostream>
#include <queue>

using namespace std;

const int MAX = 101;
int N, M;
int maze[MAX][MAX];
int dy[] = {-1, 0, 1, 0};
int dx[] = {0, 1, 0, -1};
struct XY {
  int y, x;
};
queue<XY> que;

void bfs(int y, int x) {
  que.push({y, x});
  while (!que.empty()) {
    int y = que.front().y;
    int x = que.front().x;
    que.pop();

    for (int i = 0; i < 4; i++) {
      int nextY = y + dy[i];
      int nextX = x + dx[i];

      if (nextX < 0 || nextY < 0 || nextX > M - 1 || nextY > N - 1)
        continue;
      if (maze[nextY][nextX] == 1) {
        que.push({nextY, nextX});
        maze[nextY][nextX] = maze[y][x] + 1;
      }
    }
  }
}

int main() {
  cin >> N >> M;
  for (int i = 0; i < N; i++) {
    string s;
    cin >> s;
    for (int j = 0; j < M; j++)
      maze[i][j] = s[j] - '0';
  }
  bfs(0, 0);

  cout << maze[N - 1][M - 1];
}